proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,473 @@
+
+```swift
+@propertyWrapper
+struct Clamped<Value: Comparable> {
+
+  ...
+  
+
+  init(projectedValue: Self) { ... }
+  
+}
+
+func distanceFromUpperBound(
+  @Clamped clamped: Int
+) { ... }
+
+
+distanceFromUpperBound(
+  $clamped: Clamped(to: 0 ... 100, wrappedValue: 30)
+) // returns: 70
+```
+
+
+### Add Support for `inout` Wrapped Parameters is Functions
+
+This proposal doesn't currently support marking function parameters to which wrapper types have been applied `inout`. We deemed that this functionality would be better tackled by another proposal due to its implementation complexity. However, such a feature would be really useful for wrapper types with value semantics and it would simplify the mental model. Furthermore, it could alleviate some confusion for users that don't understand the difference between a setter with value semantics and one with reference semantics.
+
+
+### Accessing Enclosing Self from Wrapper Types
+
+There's currently no public feature that allows a wrapper to access its enclosing `Self` type:
+
+```swift
+@propertyWrapper
+struct Mirror<EnclosingSelf, Value, Path>
+  where Path : KeyPath<EnclosingSelf, Value> { 
+  
+  let keyPath: Path 
+
+
+  init(of keyPath: Path) { ... }
+  
+}
+
+
+struct Point {
+
+  private var _vector: SIMD2<Double>
+  
+  
+  init(x: Double, y: Double) {
+    self._vector = SIMD2(x: x, y: y)
+  }
+  
+  
+  @Mirror(of: \._vector.x)
+  var x
+  
+  @Mirror(of: \._vector.y)
+  var y
+  
+} ❌
+// In the above use, we'd access the enclosing
+// type's '_vector' property through the provided
+// keyPath. However, today that's invalid.
+```
+
+Furthermore, extending this feature's availability to function and closure parameters be really powerful:
+
+```swift
+func valueAndIdPair<Value>(
+  @Mirror of property: Value
+) -> (value: Value, id: Int) {
+  (value: property, id: $property.keyPath.hashValue)
+}
+```
+
+It's important to note that allowing use of such a feature in function parameters would entail some limitations. For example, a parameter makred with a wrapper type referencing enclosing `Self` would not be initializable with the sugared function call that utilizes `init(wrappedValue: ...)`. That's because it would require access to the enclosing `Self`; however, in the case of calling a function this would be undefined.
+
+
+### Add Wrapper Types in the Standard Library
+
+The has been some discussion of adding wrapper types to the Standard Library, such as an `@Atomic` wrapper, so as to facilitate certain APIs. For example, `@UnsafePointer` could be a very useful one, due to the way one accesses the `pointee` property of an `UnsafePointer` type:
```

Comment:
Do you have any other ideas about wrappers we could use here? I'll look into the possibility of adding a `@Lazy` wrapper here, but other than that I can only think of `@Atomic` and `@ThreadSpecific` – which I'm not certain should be included in the Standard Library.

Reply:
Hmm, not off the top of my head but I'll keep thinking about it